File: app/src/main/java/org/mifos/mobilebanking/ui/fragments/LoanAccountTransactionFragment.java
package org.mifos.mobilebanking.ui.fragments;
/*
~This project is licensed under the open source MPL V2.
~See https://github.com/openMF/self-service-app/blob/master/LICENSE.md
*/

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import com.github.therajanmaurya.sweeterror.SweetUIErrorHandler;

import org.mifos.mobilebanking.R;
import org.mifos.mobilebanking.models.accounts.loan.LoanWithAssociations;
import org.mifos.mobilebanking.presenters.LoanAccountsTransactionPresenter;
import org.mifos.mobilebanking.ui.activities.base.BaseActivity;
import org.mifos.mobilebanking.ui.adapters.RecentTransactionListAdapter;
import org.mifos.mobilebanking.ui.fragments.base.BaseFragment;
import org.mifos.mobilebanking.ui.views.LoanAccountsTransactionView;
import org.mifos.mobilebanking.utils.Constants;
import org.mifos.mobilebanking.utils.Network;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * Created by dilpreet on 4/3/17.
 */

public class LoanAccountTransactionFragment extends BaseFragment
        implements LoanAccountsTransactionView {

    @BindView(R.id.layout_error)
    View layoutError;

    @BindView(R.id.ll_loan_account_trans)
    LinearLayout llLoanAccountTrans;

    @BindView(R.id.tv_loan_product_name)
    TextView tvLoanProductName;

    @BindView(R.id.rv_loan_transactions)
    RecyclerView rvLoanTransactions;

    @Inject
    RecentTransactionListAdapter transactionsListAdapter;

    @Inject
    LoanAccountsTransactionPresenter loanAccountsTransactionPresenter;

    private long loanId;
    private View rootView;
    private LoanWithAssociations loanWithAssociations;
    private SweetUIErrorHandler sweetUIErrorHandler;

    public static LoanAccountTransactionFragment newInstance(long loanId) {
        LoanAccountTransactionFragment fragment = new LoanAccountTransactionFragment();
        Bundle args = new Bundle();
        args.putLong(Constants.LOAN_ID, loanId);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((BaseActivity) getActivity()).getActivityComponent().inject(this);
        if (getArguments() != null) {
            loanId = getArguments().getLong(Constants.LOAN_ID);
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
            @Nullable Bundle savedInstanceState) {

        rootView = inflater.inflate(R.layout.fragment_loan_account_transactions, container, false);
        setToolbarTitle(getString(R.string.transactions));

        ButterKnife.bind(this, rootView);
        loanAccountsTransactionPresenter.attachView(this);

        sweetUIErrorHandler = new SweetUIErrorHandler(getContext(), rootView);
        showUserInterface();
        if (savedInstanceState == null) {
            loanAccountsTransactionPresenter.loadLoanAccountDetails(loanId);
        }
        return rootView;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putParcelable(Constants.LOAN_ACCOUNT, loanWithAssociations);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (savedInstanceState != null) {
            showLoanTransactions((LoanWithAssociations) savedInstanceState.
                    getParcelable(Constants.LOAN_ACCOUNT));
        }
    }

    /**
     * Initialized {@link RecyclerView} {@code rvLoanTransactions}
     */
    @Override
    public void showUserInterface() {
        LinearLayoutManager layoutManager = new LinearLayoutManager(getActivity());
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        rvLoanTransactions.setHasFixedSize(true);
        rvLoanTransactions.setLayoutManager(layoutManager);
        rvLoanTransactions.setAdapter(transactionsListAdapter);
    }

    /**
     * Fetches {@code loanWithAssociations} from server and intializes it in
     * {@code transactionsListAdapter}
     * @param loanWithAssociations object containing details about a Loan Account with Associations
     */
    @Override
    public void showLoanTransactions(LoanWithAssociations loanWithAssociations) {
        this.loanWithAssociations = loanWithAssociations;
        llLoanAccountTrans.setVisibility(View.VISIBLE);
        tvLoanProductName.setText(loanWithAssociations.getLoanProductName());
        transactionsListAdapter.setTransactions(loanWithAssociations.getTransactions());
    }

    /**
     * Sets a {@link TextView} with a msg if Transactions list is empty
     * @param loanWithAssociations
     */
    @Override
    public void showEmptyTransactions(LoanWithAssociations loanWithAssociations) {
        sweetUIErrorHandler.showSweetEmptyUI(getString(R.string.transactions),
                R.drawable.ic_compare_arrows_black_24dp, rvLoanTransactions, layoutError);
    }

    /**
     * It is called whenever any error occurs while executing a request
     * @param message Error message that tells the user about the problem.
     */
    @Override
    public void showErrorFetchingLoanAccountsDetail(String message) {
        if (!Network.isConnected(getActivity())) {
            sweetUIErrorHandler.showSweetNoInternetUI(rvLoanTransactions, layoutError);
        } else {
            sweetUIErrorHandler.showSweetErrorUI(message,
                    rvLoanTransactions, layoutError);
            Toast.makeText(getActivity(), message, Toast.LENGTH_SHORT).show();
        }
    }

    @OnClick(R.id.btn_try_again)
    public void retryClicked() {
        if (Network.isConnected(getContext())) {
            sweetUIErrorHandler.hideSweetErrorLayoutUI(rvLoanTransactions, layoutError);
            loanAccountsTransactionPresenter.loadLoanAccountDetails(loanId);
        } else {
            Toast.makeText(getContext(), getString(R.string.internet_not_connected),
                    Toast.LENGTH_SHORT).show();
        }
    }

    @Override
    public void showProgress() {
        showProgressBar();
    }

    @Override
    public void hideProgress() {
        hideProgressBar();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        hideProgressBar();
        loanAccountsTransactionPresenter.detachView();
    }
}

File: app/src/main/java/org/mifos/mobile/ui/fragments/LoanAccountTransactionFragment.kt
package org.mifos.mobile.ui.fragments

import android.os.Bundle
import android.os.Parcelable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.github.therajanmaurya.sweeterror.SweetUIErrorHandler
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import org.mifos.mobile.R
import org.mifos.mobile.databinding.FragmentLoanAccountTransactionsBinding
import org.mifos.mobile.models.accounts.loan.LoanWithAssociations
import org.mifos.mobile.ui.adapters.RecentTransactionListAdapter
import org.mifos.mobile.ui.fragments.base.BaseFragment
import org.mifos.mobile.utils.Constants
import org.mifos.mobile.utils.LoanUiState
import org.mifos.mobile.utils.Network
import org.mifos.mobile.viewModels.LoanAccountTransactionViewModel
import javax.inject.Inject

/*
~This project is licensed under the open source MPL V2.
~See https://github.com/openMF/self-service-app/blob/master/LICENSE.md
*/ /**
 * Created by dilpreet on 4/3/17.
 */
@AndroidEntryPoint
class LoanAccountTransactionFragment : BaseFragment() {
    private var _binding: FragmentLoanAccountTransactionsBinding? = null
    private val binding get() = _binding!!

    @JvmField
    @Inject
    var transactionsListAdapter: RecentTransactionListAdapter? = null

    private val viewModel: LoanAccountTransactionViewModel by viewModels()

    private var loanId: Long? = 0
    private var loanWithAssociations: LoanWithAssociations? = null
    private var sweetUIErrorHandler: SweetUIErrorHandler? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (arguments != null) {
            loanId = arguments?.getLong(Constants.LOAN_ID)
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        _binding = FragmentLoanAccountTransactionsBinding.inflate(inflater, container, false)
        val rootView = binding.root
        setToolbarTitle(getString(R.string.transactions))
        sweetUIErrorHandler = SweetUIErrorHandler(context, rootView)
        showUserInterface()
        if (savedInstanceState == null) {
            viewModel.loadLoanAccountDetails(loanId)
        }
        return rootView
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putParcelable(Constants.LOAN_ACCOUNT, loanWithAssociations)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        if (savedInstanceState != null) {
            showLoanTransactions(savedInstanceState.getParcelable<Parcelable>(Constants.LOAN_ACCOUNT) as LoanWithAssociations)
        }
    }

    /**
     * Initialized [RecyclerView] `rvLoanTransactions`
     */
    fun showUserInterface() {
        val layoutManager = LinearLayoutManager(activity)
        layoutManager.orientation = LinearLayoutManager.VERTICAL
        with(binding) {
            rvLoanTransactions.setHasFixedSize(true)
            rvLoanTransactions.layoutManager = layoutManager
            rvLoanTransactions.adapter = transactionsListAdapter
        }
    }

    /**
     * Fetches `loanWithAssociations` from server and intializes it in
     * `transactionsListAdapter`
     *
     * @param loanWithAssociations object containing details about a Loan Account with Associations
     */
    fun showLoanTransactions(loanWithAssociations: LoanWithAssociations?) {
        this.loanWithAssociations = loanWithAssociations
        binding.llLoanAccountTrans.visibility = View.VISIBLE
        binding.tvLoanProductName.text = loanWithAssociations?.loanProductName
        transactionsListAdapter?.setTransactions(loanWithAssociations?.transactions)
    }

    /**
     * Sets a [TextView] with a msg if Transactions list is empty
     */
    fun showEmptyTransactions(loanWithAssociations: LoanWithAssociations?) {
        sweetUIErrorHandler?.showSweetEmptyUI(
            getString(R.string.transactions),
            R.drawable.ic_compare_arrows_black_24dp,
            binding.rvLoanTransactions,
            binding.layoutError.root,
        )
    }

    /**
     * It is called whenever any error occurs while executing a request
     *
     * @param message Error message that tells the user about the problem.
     */
    fun showErrorFetchingLoanAccountsDetail(message: String?) {
        with(binding) {
            if (!Network.isConnected(activity)) {
                sweetUIErrorHandler?.showSweetNoInternetUI(
                    rvLoanTransactions,
                    layoutError.root,
                )
            } else {
                sweetUIErrorHandler?.showSweetErrorUI(
                    message,
                    rvLoanTransactions,
                    layoutError.root,
                )
                Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        viewLifecycleOwner.lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                viewModel.loanUiState.collect {
                    when (it) {
                        is LoanUiState.Loading -> showProgress()
                        is LoanUiState.ShowError -> {
                            hideProgress()
                            showErrorFetchingLoanAccountsDetail(getString(it.message))
                        }
                        is LoanUiState.ShowLoan -> {
                            hideProgress()
                            showLoanTransactions(it.loanWithAssociations)
                        }
                        is LoanUiState.ShowEmpty -> {
                            hideProgress()
                            showEmptyTransactions(it.loanWithAssociations)
                        }
                        else -> throw IllegalStateException("Unexpected state: $it")
                    }
                }
            }
        }

        binding.layoutError.btnTryAgain.setOnClickListener {
            retryClicked()
        }
    }

    fun retryClicked() {
        if (Network.isConnected(context)) {
            sweetUIErrorHandler?.hideSweetErrorLayoutUI(
                binding.rvLoanTransactions,
                binding.layoutError.root,
            )
            viewModel.loadLoanAccountDetails(loanId)
        } else {
            Toast.makeText(
                context,
                getString(R.string.internet_not_connected),
                Toast.LENGTH_SHORT,
            ).show()
        }
    }

    fun showProgress() {
        showProgressBar()
    }

    fun hideProgress() {
        hideProgressBar()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        hideProgressBar()
        _binding = null
    }

    companion object {
        fun newInstance(loanId: Long?): LoanAccountTransactionFragment {
            val fragment = LoanAccountTransactionFragment()
            val args = Bundle()
            if (loanId != null) args.putLong(Constants.LOAN_ID, loanId)
            fragment.arguments = args
            return fragment
        }
    }
}
